### PR TITLE
cross-platform: x86 support

### PR DESCRIPTION
**Pull request by obastemur**
ChakraCore compiles to x86 target

- Add xplat x86 Support [~~not tested - linker problems - see below~~]
- Improvements to existing amd64 implementation
- ~~This PR needs 'ICU/etc static linkage' support for linker pass~~
- Additional work will be needed on posix (At the moment, the main focus for this x86 PR is iOS simulator / Android targets)

Important: Unless otherwise specified, x86 is not officially supported.

### CODE REVIEW

**Comment by obastemur** [comment]
At the moment, this PR is capable of running javascript on x86 arch / unix. There are couple of missing parts (i.e. function call  stack failures)

Since this is not an official support and this work is beyond compiling step now; (also the improvements to 64 bit side)
@digitalinfinity @jianchun how about merging this work as is (apart from the review issues) and handling the rest in smaller iterations ?

File: lib/Common/Common/NumberUtilities.inl [hunk 1]
```diff
@@ -72,7 +72,7 @@ namespace Js
     NUMBER_UTIL_INLINE INT64 NumberUtilities::TryToInt64(double T1)
     {
         INT64 T4_64;
-#if defined(_M_IX86)
+#if defined(_M_IX86) && defined(_WIN32) // TODO: xplat support
```
Comment (digitalinfinity) [inline lib/Common/Common/NumberUtilities.inl]: Why not do the same thing you did for AutoSystemInfo::Data::SSE2Available, and just have that function return false with a todo

File: lib/Common/Memory/HeapBlock.cpp [hunk 2]
```diff
@@ -1590,7 +1590,7 @@ void SmallHeapBlockT<TBlockAttributes>::FillFreeMemory(__in_bcount(size) void *
     }
 
     // REVIEW: Do DbgMemFill on debug build?
-#if defined(_M_IX86)
+#if defined(_M_IX86) && defined(_WIN32) // TODO: xplat support
```
Comment (digitalinfinity) [inline lib/Common/Memory/HeapBlock.cpp]: Why is this _WIN32 specific?

File: lib/Common/Common/NumberUtilities.inl (hunk 1 above)
Comment (obastemur) [inline lib/Common/Common/NumberUtilities.inl]: Because the instructions are not available on xplat. So, it doesn't compile

Comment (digitalinfinity) [inline lib/Common/Common/NumberUtilities.inl]: I'm confused- FLD and FISTTP are not understood by other x86 assemblers? Or is it that we need to pass in a switch to get them to understand x87 instructions? Because FLD is a very old instruction. FISTTP is relatively newer- I think -msse3 should support both. Maybe it should get wrapped in a different ifdef (something like _M_CHAKRA_SSE3 ifdef that the makefile can pass in along with -msse3 to the compiler?)

---
In reply to: [78863793](https://github.com/Microsoft/ChakraCore/pull/1540#discussion_r78863793) [](ancestors = 78863793)

File: lib/Runtime/Library/JavascriptFunction.cpp [hunk 3]
```diff
@@ -1126,8 +1126,24 @@ namespace Js
...
+{
+    // todo: fix this!!!
+    __asm
+    {
+        push    ecx
+        lea     ecx, [esp + 8]
+        sub     ecx, eax
+        and     ecx, (16 - 1)
+        add     eax, ecx
+        ret
```
Comment (digitalinfinity) [inline lib/Runtime/Library/JavascriptFunction.cpp]: Confused by what this is doing but I guess that's why there is a todo. I'm assuming the fix involves changing esp. Right now, it looks like this would leave the stack unaligned since you do a push ecx but no pop?

File: lib/Runtime/Library/JavascriptNumber.cpp [hunk 4]
```diff
@@ -230,6 +230,14 @@ namespace Js
 
     static const double d1_0 = 1.0;
 
+#ifdef __clang__
+    // todo : make this fast!
+    // Clang doesn't support so many things under the fast version below
```
Comment (digitalinfinity) [inline lib/Runtime/Library/JavascriptNumber.cpp]: Weird- are we using -msse3?

File: lib/Runtime/Library/MathLibrary.cpp [hunk 5]
```diff
@@ -147,7 +147,7 @@ namespace Js
     {
         double result;
 
-#if defined(_M_IX86)
+#if defined(_M_IX86) && defined(_WIN32) // TODO: xplat support
```
Comment (digitalinfinity) [inline lib/Runtime/Library/MathLibrary.cpp]: Yeah, I'd do this under a different ifdef (assuming -msse3 doesn't fix it)

File: pal/inc/pal.h [hunk 6]
```diff
@@ -2026,7 +2026,7 @@ QueueUserAPC(
          IN HANDLE hThread,
          IN ULONG_PTR dwData);
 
-#ifdef _X86_
+#ifdef __i686__
```
Comment (digitalinfinity) [inline pal/inc/pal.h]: What's the reason for this change?

File: pal/inc/pal_mstypes.h [hunk 7]
```diff
@@ -42,22 +42,11 @@ extern "C" {
...
-#if !defined(__cdecl)
-#define __cdecl        __attribute__((cdecl))
-#endif
-#if !defined(_cdecl)
-#define _cdecl         __cdecl
-#endif
-#if !defined(CDECL)
-#define CDECL          __cdecl
-#endif
+#define __stdcall
```
Comment (digitalinfinity) [inline pal/inc/pal_mstypes.h]: Curious- does clang not support stdcall?

File: pal/src/arch/i386/activationhandlerwrapper.S [hunk 8]
```diff
@@ -28,4 +28,23 @@ LOCAL_LABEL(ActivationHandlerReturn):
     ret
 NESTED_END ActivationHandlerWrapper, _TEXT
 
+#else
+
+// Offset of the return address from the ActivationHandler in the ActivationHandlerWrapper
+.globl C_FUNC(ActivationHandlerReturnOffset)
```
Comment (digitalinfinity) [inline pal/src/arch/i386/activationhandlerwrapper.S]: Who needs this?

File: pal/src/debug/debug.cpp [hunk 9]
```diff
@@ -422,7 +422,12 @@ DebugBreak(
 BOOL
 IsInDebugBreak(void *addr)
 {
+#ifdef _M_IX86
+    // todo: implement this
+    return FALSE;
```
Comment (digitalinfinity) [inline pal/src/debug/debug.cpp]: Hmm- why does the original implementation not work?

**Comment by digitalinfinity** [comment]
:shipit:

File: pal/src/debug/debug.cpp (hunk 9 above)
Comment (obastemur) [inline pal/src/debug/debug.cpp]: Depending PAL part fails. Disabled temporarily

File: pal/src/arch/i386/activationhandlerwrapper.S (hunk 8 above)
Comment (obastemur) [inline pal/src/arch/i386/activationhandlerwrapper.S]: machexception

File: pal/inc/pal_mstypes.h (hunk 7 above)
Comment (obastemur) [inline pal/inc/pal_mstypes.h]: we already disabled it for x64. It is not because it doesn't support that. It is because of our usage of templates.

File: lib/Runtime/Library/JavascriptFunction.cpp (hunk 3 above)
Comment (obastemur) [inline lib/Runtime/Library/JavascriptFunction.cpp]: close. there is much more to do there (i.e. we use more than 0x1000 ??)

File: lib/Runtime/Library/MathLibrary.cpp (hunk 5 above)
Comment (obastemur) [inline lib/Runtime/Library/MathLibrary.cpp]: @digitalinfinity `__libm_sse2_acos` (msvcrt) is the problem there.

File: lib/Runtime/Library/JavascriptNumber.cpp (hunk 4 above)
Comment (obastemur) [inline lib/Runtime/Library/JavascriptNumber.cpp]: It is not about msse3 etc. Clang inline assembly does not support non-asm in naked function.

File: lib/Runtime/Library/JavascriptFunction.cpp (hunk 3 above)
Comment (curtisman) [inline lib/Runtime/Library/JavascriptFunction.cpp]: So this doesn't really work at all? since it doesn't move ESP?
Should we just do "int 3" here so we are guarantee to crash?

Comment (obastemur) [inline lib/Runtime/Library/JavascriptFunction.cpp]: `int 3` is a good idea

File: lib/Runtime/Library/JavascriptNumber.cpp (hunk 4 above)
Comment (curtisman) [inline lib/Runtime/Library/JavascriptNumber.cpp]: I think this is only called by the JIT.  And this change will break that.
Might want to put it under #ifdef ENABLE_NATIVE_CODEGEN instead?  So we will hit this and fix it property when we work on x86 JIT?

File: pal/inc/pal_mstypes.h (hunk 7 above)
Comment (curtisman) [inline pal/inc/pal_mstypes.h]: So template doesn't support stdcall?

Comment (obastemur) [inline pal/inc/pal_mstypes.h]: Surely supports but we do violate some things clang is pretty strict

File: pal/src/debug/debug.cpp (hunk 9 above)
Comment (curtisman) [inline pal/src/debug/debug.cpp]: Put more detail in the comment please.

File: lib/Runtime/Library/MathLibrary.cpp (hunk 5 above)
Comment (curtisman) [inline lib/Runtime/Library/MathLibrary.cpp]: Yeah... this depends on whether the xplat CRT has a direct entry point to the SSE2 version of these trig functions.....  My guess is probably not :(  Even if they have it is probably under a different name...  defined(_WIN32) is probably fine.
A better comment would be great.

Comment (obastemur) [inline lib/Runtime/Library/MathLibrary.cpp]: @curtisman  Thought `lib_sse2_acos` can be implemented. So no need ?

Comment (curtisman) [inline lib/Runtime/Library/MathLibrary.cpp]: That probably isn't necessary. This is really for perf.  Functionally, this is not necessary.
If the xplat CRT implementation already support SSE2, then we get most of the perf already.
It just avoid the is SSE2 support check at the beginning of these functions.

I'd rather not maintain our copy of these trig functions.

Comment (obastemur) [inline lib/Runtime/Library/MathLibrary.cpp]: updated accordingly

**Comment by obastemur** [comment]
@digitalinfinity @curtisman Thanks for the review
